### PR TITLE
Allow checker to support reparsepoints in Windows

### PR DESCRIPTION
**Pull request by fdncred**

### CODE REVIEW

**Comment by jonathanpallant** [comment]
On Linux, `symlink_meta` has `metadata.is_file == False` for symlinks, which means this crate won't find symlinks in /usr/bin for example.

```
Meta for "/usr/bin/xzcat":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                mode: 33261,
            },
        ),
        is_dir: false,
        is_file: true,
        permissions: Permissions(
            FilePermissions {
                mode: 33261,
            },
        ),
        modified: Ok(
            SystemTime {
                tv_sec: 1498667959,
                tv_nsec: 0,
            },
        ),
        accessed: Ok(
            SystemTime {
                tv_sec: 1591012305,
                tv_nsec: 0,
            },
        ),
        created: Ok(
            SystemTime {
                tv_sec: 1591012379,
                tv_nsec: 420000000,
            },
        ),
    },
)
Symlink Meta for "/usr/bin/xzcat":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                mode: 41471,
            },
        ),
        is_dir: false,
        is_file: false,
        permissions: Permissions(
            FilePermissions {
                mode: 41471,
            },
        ),
        modified: Ok(
            SystemTime {
                tv_sec: 1498667959,
                tv_nsec: 0,
            },
        ),
        accessed: Ok(
            SystemTime {
                tv_sec: 1592493357,
                tv_nsec: 230000000,
            },
        ),
        created: Ok(
            SystemTime {
                tv_sec: 1591012379,
                tv_nsec: 420000000,
            },
        ),
    },
)
Got Output { status: ExitStatus(ExitStatus(0)), stdout: "xz (XZ Utils) 5.2.2\nliblzma 5.2.2\n", stderr: "" }
Meta for "/usr/bin/xz":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                mode: 33261,
            },
        ),
        is_dir: false,
        is_file: true,
        permissions: Permissions(
            FilePermissions {
                mode: 33261,
            },
        ),
        modified: Ok(
            SystemTime {
                tv_sec: 1498667959,
                tv_nsec: 0,
            },
        ),
        accessed: Ok(
            SystemTime {
                tv_sec: 1592493388,
                tv_nsec: 410000000,
            },
        ),
        created: Ok(
            SystemTime {
                tv_sec: 1591012379,
                tv_nsec: 420000000,
            },
        ),
    },
)
Symlink Meta for "/usr/bin/xz":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                mode: 33261,
            },
        ),
        is_dir: false,
        is_file: true,
        permissions: Permissions(
            FilePermissions {
                mode: 33261,
            },
        ),
        modified: Ok(
            SystemTime {
                tv_sec: 1498667959,
                tv_nsec: 0,
            },
        ),
        accessed: Ok(
            SystemTime {
                tv_sec: 1592493388,
                tv_nsec: 410000000,
            },
        ),
        created: Ok(
            SystemTime {
                tv_sec: 1591012379,
                tv_nsec: 420000000,
            },
        ),
    },
)
Got Output { status: ExitStatus(ExitStatus(0)), stdout: "xz (XZ Utils) 5.2.2\nliblzma 5.2.2\n", stderr: "" }
```

```
❯ cargo run
   Compiling file_test v0.1.0 (C:\Users\jpallant\file_test)
    Finished dev [unoptimized + debuginfo] target(s) in 0.54s
     Running `target\debug\file_test.exe`
Meta for "C:\\Users\\jpallant\\AppData\\Local\\Microsoft\\WindowsApps\\winget.exe":
Err(
    Os {
        code: 1920,
        kind: Other,
        message: "The file cannot be accessed by the system.",
    },
)
Symlink Meta for "C:\\Users\\jpallant\\AppData\\Local\\Microsoft\\WindowsApps\\winget.exe":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                attributes: 1056,
                reparse_tag: 2147483675,
            },
        ),
        is_dir: false,
        is_file: true,
        permissions: Permissions(
            FilePermissions {
                attrs: 1056,
            },
        ),
        modified: Ok(
            SystemTime {
                intervals: 132354869816004065,
            },
        ),
        accessed: Ok(
            SystemTime {
                intervals: 132354869816004065,
            },
        ),
        created: Ok(
            SystemTime {
                intervals: 132354869816004065,
            },
        ),
    },
)
Got Output { status: ExitStatus(ExitStatus(0)), stdout: "v0.1.41331 Preview", stderr: "" }
Meta for "C:\\Python38\\Python.exe":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                attributes: 32,
                reparse_tag: 0,
            },
        ),
        is_dir: false,
        is_file: true,
        permissions: Permissions(
            FilePermissions {
                attrs: 32,
            },
        ),
        modified: Ok(
            SystemTime {
                intervals: 132338797780000000,
            },
        ),
        accessed: Ok(
            SystemTime {
                intervals: 132369667820021804,
            },
        ),
        created: Ok(
            SystemTime {
                intervals: 132338797780000000,
            },
        ),
    },
)
Symlink Meta for "C:\\Python38\\Python.exe":
Ok(
    Metadata {
        file_type: FileType(
            FileType {
                attributes: 32,
                reparse_tag: 0,
            },
        ),
        is_dir: false,
        is_file: true,
        permissions: Permissions(
            FilePermissions {
                attrs: 32,
            },
        ),
        modified: Ok(
            SystemTime {
                intervals: 132338797780000000,
            },
        ),
        accessed: Ok(
            SystemTime {
                intervals: 132369667820021804,
            },
        ),
        created: Ok(
            SystemTime {
                intervals: 132338797780000000,
            },
        ),
    },
)
Got Output { status: ExitStatus(ExitStatus(0)), stdout: "Python 3.8.3\r\n", stderr: "" }
```


```rust
use std::process::Command;

#[cfg(target_os="windows")]
static PATHS: [&str; 2] = [
    r"C:\Users\jpallant\AppData\Local\Microsoft\WindowsApps\winget.exe",
    r"C:\Python38\Python.exe",
];

#[cfg(target_os="linux")]
static PATHS: [&str; 2] = [
    r"/usr/bin/xzcat",
    r"/usr/bin/xz",
];

fn main() {
    for path in &PATHS {
        println!("Meta for {:?}:\n{:#?}", path, std::fs::metadata(path));
        println!(
            "Symlink Meta for {:?}:\n{:#?}",
            path,
            std::fs::symlink_metadata(path)
        );

        let output = Command::new(path)
            .arg("--version")
            .output()
            .expect("Failed to execute command");

        println!("Got {:?}", output);
    }
}
```

**Comment by harryfei** [comment]
Does that mean we should use different `ExistedChecker` implementations for Linux and Windows?

**Comment by jonathanpallant** [comment]
Yeah I can't see a way around that at the moment. Windows' concept of symlinks is quite different to POSIX OSes like Linux and macOS.

**Comment by fdncred** [comment]
ok, @jonathanpallant. Thanks for testing. @harryfei , how does this?

```rust
impl Checker for ExistedChecker {
    #[cfg(target_os="windows")]
    fn is_valid(&self, path: &Path) -> bool {
        fs::symlink_metadata(path)
            .map(|metadata| metadata.is_file())
            .unwrap_or(false)
    }

    #[cfg(not(target_os="windows"))]
    fn is_valid(&self, path: &Path) -> bool {
        fs::metadata(path)
            .map(|metadata| metadata.is_file())
            .unwrap_or(false)
    }
}
```

**Comment by fdncred** [comment]
@harryfei Any idea what this CI error means?

**Comment by fdncred** [comment]
CI is glitchy here. Not sure what needs to happen to get this merged. @harryfei can you help please?

**Comment by harryfei** [comment]
So odd. I can merge it without passing all the CI jobs. 
I think you will delete the `Tweak to get CI to run again. ` commit.

**Comment by fdncred** [comment]
@harryfei Deleted the `Tweak` commit. Looks like all checks finally passed. I have no idea what was causing that glitch.

File: src/checker.rs [hunk 1]
```diff
@@ -23,7 +23,7 @@ impl Checker for ExecutableChecker {
             .and_then(|c| Ok(unsafe { libc::access(c.as_ptr(), libc::X_OK) == 0 }))
             .unwrap_or(false)
     }
-
+
```
Comment (harryfei) [inline src/checker.rs:26]: unexpected spaces.

Comment (fdncred) [reply src/checker.rs:26]: are you saying you don't want these spaces or i should remove the line entirely?

**Pull request merged**